test_conformance/extensions/cl_khr_command_buffer/command_buffer_printf.cpp

```diff
@@ -0,0 +1,531 @@
+struct CommandBufferPrintfTest : public BasicCommandBufferTest
+{
+    CommandBufferPrintfTest(cl_device_id device, cl_context context,
+                            cl_command_queue queue)
+        : BasicCommandBufferTest(device, context, queue),
+          trigger_event(nullptr), wait_event(nullptr), file_descriptor(0),
+          printf_use_support(false)
+    {
+        simultaneous_use_requested = simul_use;
+        if (simul_use)
+        {
+            buffer_size_multiplier = num_test_iters;
+        }
+    }
+
+    //--------------------------------------------------------------------------
+    void ReleaseOutputStream(int fd)
+    {
+        fflush(stdout);
+        streamDup2(fd, fileno(stdout));
+        close(fd);
+    }
+
+    //--------------------------------------------------------------------------
+    int AcquireOutputStream(int* error)
+    {
+        int fd = streamDup(fileno(stdout));
+        *error = 0;
+        if (!freopen(temp_filename.c_str(), "wt", stdout))
+        {
+            ReleaseOutputStream(fd);
+            *error = -1;
+        }
+        return fd;
+    }
+
+    //--------------------------------------------------------------------------
+    void GetAnalysisBuffer(std::stringstream& buffer)
+    {
+        std::ifstream fp(temp_filename, std::ios::in);
+        if (fp.is_open())
+        {
+            buffer << fp.rdbuf();
+        }
+    }
+
+    //--------------------------------------------------------------------------
+    void PurgeTempFile()
+    {
+        std::ofstream ofs(temp_filename,
+                          std::ofstream::out | std::ofstream::trunc);
+        ofs.close();
+    }
+
+    //--------------------------------------------------------------------------
+    bool Skip() override
+    {
+        // Query if device supports kernel printf use
+        cl_device_command_buffer_capabilities_khr capabilities;
+        cl_int error =
+            clGetDeviceInfo(device, CL_DEVICE_COMMAND_BUFFER_CAPABILITIES_KHR,
+                            sizeof(capabilities), &capabilities, NULL);
+        test_error(error,
+                   "Unable to query CL_DEVICE_COMMAND_BUFFER_CAPABILITIES_KHR");
+
+        printf_use_support =
+            (capabilities & CL_COMMAND_BUFFER_CAPABILITY_KERNEL_PRINTF_KHR)
+            != 0;
+
+        if (!printf_use_support) return true;
+        return BasicCommandBufferTest::Skip()
+            || (simultaneous_use_requested && !simultaneous_use_support);
+    }
+
+    //--------------------------------------------------------------------------
+    cl_int SetUpKernel() override
+    {
+        cl_int error = CL_SUCCESS;
+
+        const char* kernel_str =
+            R"(
+      __kernel void print(__global char* in, __global char* out, __global int* offset)
+      {
+          size_t id = get_global_id(0);
+          int ind = offset[0] + offset[1] * id;
+          for(int i=0; i<offset[1]; i++) out[ind+i] = in[i];
+          printf("%s", in);
```

Comment:
Hello Ben

Thanks for your feedback, I appreciate it.

Please let me summarize the idea behind that kernel. The goal was to print 1-6 the same characters in a null terminated string. I am using `std::vector<cl_char>` which guarantee consistent buffer (which is not that obvious when it comes to `std::string`). Then, mentioned vector is copied into `cl_mem` object which is set as first argument of kernel object:

`__global char* in`

next, `in` is used as `printf` argument in kernel call.

Could you explain a bit more what correction is on your mind ? Thank you

Reply:
I think the issue is that `__global char* in` isn't a string literal, can see in the spec section on examples of invalid usage of `%s` we've fallen into the first case :

![image](https://user-images.githubusercontent.com/2334151/217848137-1bb1d36f-0190-4877-801d-0bcc0590aacd.png)

Reply:
ops :o, good find. Can't find clear info in spec - will below correction be sufficient? Guess not.  

`__global const char* in`

In that case we will have to "bake" string literal into kernel source, right ? Other ideas ?

Reply:
Using `%c` as per your example seems like the easiest fix to the immediate problem of invalid Opencl-C to me.

Reply:
Done

Reply:
Thanks!  The OpenCL C program looks valid now and this test is now passing on the Intel implementation with the command buffer emulation layer :+1: .

Interestingly the test is still failing on a different implementation.  I haven't been able to figure out why but I don't think it's related to command buffers or the command buffer emulation layer, since the test also fails if I replace the call to `clEnqueueCommandBufferKHR` with a call to `clEnqueueNDRangeKernel`:

```c++
#if 0
        // enqueue command buffer with kernel containing printf command
        error = clEnqueueCommandBufferKHR(0, nullptr, command_buffer, 0,
                                          nullptr, &wait_event);
        test_error_release_stdout(error, "clEnqueueCommandBufferKHR failed");
#else
        error = clEnqueueNDRangeKernel(queue, kernel, 1, nullptr, &num_elements, nullptr, 0, nullptr, &wait_event);
        test_error_release_stdout(error, "clEnqueueNDRangeKernel failed");
#endif
```

I'm going to resolve this conversation since this looks reasonable to me now, but this might be something to keep an eye on...

Reply:
Interestingly, after the `%c` change I now see the `basic_printf` test fail on our implementation. This is because too few characters are printed on the 3rd & last iteration of `RunSingle`. The number of total characters printed by this point should still be less than [CL_DEVICE_PRINTF_BUFFER_SIZE](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#CL_DEVICE_PRINTF_BUFFER_SIZE), so I think this is a bug in our implementation. As we've currently implemented printf in command-buffers, every enqueue of a command-buffer will use up more of the same kernel printf buffer, rather than having the whole allocation each command-buffer enqueue.

However, since another implementation is also failing, and this issue wasn't there with `%s` I thought I'd point out that I could work around this issue by using a smaller global size in the kernel, e.g instead of `num_elements` (which is `16384`) I used `1024` work-items.